src/libraries/System.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/JSSynchronizationContext.cs

```diff
@@ -41,38 +43,116 @@ public WorkItem(SendOrPostCallback callback, object? data, ManualResetEventSlim?
             }
         }
 
-        public JSSynchronizationContext(bool isMainThread)
+        // this need to be called from JSWebWorker or UI thread
+        public static JSSynchronizationContext InstallWebWorkerInterop(bool isMainThread, CancellationToken cancellationToken)
+        {
+            var ctx = new JSSynchronizationContext(isMainThread, cancellationToken);
+            ctx.previousSynchronizationContext = SynchronizationContext.Current;
+            SynchronizationContext.SetSynchronizationContext(ctx);
+
+            var proxyContext = ctx.ProxyContext;
+            JSProxyContext.CurrentThreadContext = proxyContext;
+            JSProxyContext.ExecutionContext = proxyContext;
+            if (isMainThread)
+            {
+                JSProxyContext.MainThreadContext = proxyContext;
+            }
+
+            ctx.AwaitNewData();
+
+            Interop.Runtime.InstallWebWorkerInterop(proxyContext.ContextHandle);
+
+            return ctx;
+        }
+
+        // this need to be called from JSWebWorker thread
+        internal void UninstallWebWorkerInterop()
+        {
+            if (_isDisposed)
+            {
+                return;
+            }
+            if (!_isRunning)
+            {
+                return;
+            }
+
+            var jsProxyContext = JSProxyContext.AssertIsInteropThread();
+            if (jsProxyContext != ProxyContext)
+            {
+                Environment.FailFast($"UninstallWebWorkerInterop failed, ManagedThreadId: {Environment.CurrentManagedThreadId}. {Environment.NewLine} {Environment.StackTrace}");
+            }
+            if (SynchronizationContext.Current == this)
+            {
+                SynchronizationContext.SetSynchronizationContext(this.previousSynchronizationContext);
+            }
+
+            // this will runtimeKeepalivePop()
+            // and later maybeExit() -> __emscripten_thread_exit()
+            jsProxyContext.Dispose();
+
+            JSProxyContext.CurrentThreadContext = null;
+            JSProxyContext.ExecutionContext = null;
+            _isRunning = false;
+            Dispose();
+        }
+
+        public JSSynchronizationContext(bool isMainThread, CancellationToken cancellationToken)
         {
             ProxyContext = new JSProxyContext(isMainThread, this);
             Queue = Channel.CreateUnbounded<WorkItem>(new UnboundedChannelOptions { SingleWriter = false, SingleReader = true, AllowSynchronousContinuations = true });
-            _DataIsAvailable = DataIsAvailable;
+            _ScheduleJSPump = ScheduleJSPump;
+
+            // receive callback (on any thread) that cancelation is requested
+            _cancellationTokenRegistration = cancellationToken.Register(() =>
+            {
+
+                _isCancellationRequested = true;
+                Queue.Writer.TryComplete();
+
+                while (Queue.Reader.TryRead(out var item))
+                {
+                    item.Signal?.Set();
```

Comment:
Can you explain the scenario this is meant for? I'm not sure if signaling work items that haven't actually run is correct, even if the goal is to ensure we don't hang

Reply:
There is test for `_isCancellationRequested` after `signal.Wait();` in `Send`